[torchbind] Add generic __deepcopy__ method

angelayi opened the pull request:
Summary: Added a generic `__deepcopy__` method which will use the torchbind object's existing `__getattr__` and `__setattr__` to copy the torchbind object. This will later be used in [D64124825](https://www.internalfb.com/diff/D64124825)


Differential Revision: D64124826

zou3519 commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -866,6 +866,39 @@ void initJitScriptBindings(PyObject* module) {
                 return std::hash<c10::ivalue::Object*>{}(self._ivalue().get());
               })
+          .def(
+              "_clone",
```
Why does this need to be implemented in C++? We could have a deepcopy(torchbind_obj) in Python that literally calls unflatten/flatten

angelayi replied:
I'm not sure how to attach a __deepcopy__ implementation to the torchbind python object, so I implemented it here. (updated it to directly implement __deepcopy__ instead of _clone)

zou3519 replied:
What I meant was we could have a "torch.utils.torchbind.deepcopy(obj)" (or something) that we invoke in AOTAutograd to do the copy. Though it's not clear to me which is better.

The other things I'm not sure about are:
- Is it guaranteed that flatten + unflatten on a real torchbind object gets us the correct behavior? This might be true for our test cases, but I'm not sure if we require this to be true in general

angelayi replied:
> What I meant was we could have a "torch.utils.torchbind.deepcopy(obj)" (or something) that we invoke in AOTAutograd to do the copy. Though it's not clear to me which is better.

I was thinking maybe implementing `__deepcopy__` might be better as we can also use it to copy the inputs and run with the first custom op pass (example in D64124825).

> Is it guaranteed that flatten + unflatten on a real torchbind object gets us the correct behavior? This might be true for our test cases, but I'm not sure if we require this to be true in general.

Yeah I guess it might be the case that the user's flatten/unflatten calls are not actually copying... for example the original tensor queue flatten implementation did not do a copy of the queue's contents. Is there a way to require this to be true? 

> Real tensor prop implies that we shouldn't need any user intervention. Requiring the user to define unflatten/flatten is user intervention. Is there anything else we can do here?

User intervention is only necessary if the operation results in a mutation on the obj, which I guess is somewhat common since the goal of the torchbind obj is to store state.. I'm not sure what's another thing we could do to preserve the mutations...

angelayi replied:
Updated to just implement a `__deepcopy__`

angelayi committed:
[torchbind] Add generic _clone method (#137613)

Summary:

Added a generic `torchbind_obj._clone()` which will clone the torchbind object. This requires that a `__obj_flatten__` and `__obj_unflatten__` has been implemented in cpp.

Test Plan: `buck2 run @//mode/dev-nosan //caffe2/test:jit -- -r "TestTorchbind.test_clone"`

Differential Revision: D64124826

ydwu4 commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -866,6 +866,53 @@ void initJitScriptBindings(PyObject* module) {
                 // Similar to Tensor's `__hash__`, which is `id()`.
                 return std::hash<c10::ivalue::Object*>{}(self._ivalue().get());
               })
+          .def(
```
Re-using the pickling infra for torch bind objects looks good to me since all existing usage of script objects supports serde. Can we somehow re-use the two functions defined for pickling to reduce some redundant code?

wdvr commented:
@pytorchbot merge
